subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/ZipHasher.java

```diff
@@ -110,22 +142,116 @@ private void fingerprintZipEntries(String parentName, List<FileSystemLocationFin
             }
             String fullName = parentName.isEmpty() ? zipEntry.getName() : parentName + "/" + zipEntry.getName();
             if (isZipFile(zipEntry.getName())) {
-                fingerprintZipEntries(fullName, fingerprints, new StreamZipInput(zipEntry.getInputStream()));
+                fingerprintZipEntries(fullName, rootParentName, fingerprints, new StreamZipInput(zipEntry.getInputStream()));
+            } else if (isManifestFile(zipEntry.getName())) {
+                fingerprintZipEntryContentWithFallback(rootParentName, fingerprints, zipEntry, fullName, this::hashManifest);
+            } else if (isManifestPropertyFile(zipEntry.getName())) {
+                fingerprintZipEntryContentWithFallback(rootParentName, fingerprints, zipEntry, fullName, this::hashProperties);
             } else {
-                HashCode hash = resourceHasher.hash(zipEntry);
-                if (hash != null) {
-                    fingerprints.add(new DefaultFileSystemLocationFingerprint(fullName, FileType.RegularFile, hash));
+                fingerprintZipEntry(zipEntry, fullName, fingerprints);
+            }
+        }
+    }
+
+    private void fingerprintZipEntry(ZipEntry zipEntry, String fullName, List<FileSystemLocationFingerprint> fingerprints) throws IOException {
+        HashCode hash = resourceHasher.hash(zipEntry);
+        if (hash != null) {
+            fingerprints.add(new DefaultFileSystemLocationFingerprint(fullName, FileType.RegularFile, hash));
+        }
+    }
+
+    private void fingerprintZipEntryContentWithFallback(String rootParentName, List<FileSystemLocationFingerprint> fingerprints, ZipEntry zipEntry, String fullName, ByteContentHasher hasher) throws IOException {
+        // JdkZipEntry can be backed by a stream, so we assume that getInputStream is a single shot and read the manifest to a byte array so we can fallback should content hashing fail
+        byte[] entryBytes = ByteStreams.toByteArray(zipEntry.getInputStream());
+        try {
+            HashCode hash = hasher.hash(entryBytes);
+            if (hash != null) {
+                fingerprints.add(new DefaultFileSystemLocationFingerprint(fullName, FileType.RegularFile, hash));
+            }
+        } catch (Exception e) {
+            LOGGER.warn("Could not load fingerprint " + rootParentName + ". Falling back to regular fingerprinting", e);
```

Comment:
Both `java.util.jar.Manifest` and `java.util.Properties` will throw on malformed files and instead of duplicating a fallback in `hashManifest` and `hashAttributes` I added the fallback in this outer method.

I personally want the warning, or at least very obvious output from the build cache debug logging about what's doing on, because problems with fingerprinting have been impossible to diagnose without attaching a debugger in my experience.

IIRC `rootParentName` is required because this recursively hashes zips within zips and the filename becomes part of the path.

The method takes a `ByteContentHasher`, falling back to `fingerprintZipEntry` on an exception which is why I called it`fingerprintZipEntryContentWithFallback`. `hashAttributes` and `hashProperties` normalize the file contents, and happens to filter entries while it's at it.

Reply:
What I'm saying is we don't need a fallback at all: if your properties file or manifest is malformed, the build should fail. Ignoring attributes and properties should not be a best effort option: if you want to use it, you need to make sure your files conform to the requirements. Consistency, debuggability and discoverability are the goals here, not to make as many builds as possible pass.

Once we remove the fallback, we can remove the rest I mentioned.

I'm not sure about the implications of adding `rootParentName`, but it sounds like an orthogonal issue with recursive ZIP hashing. Let's address that in a separate PR if you want to.

Reply:
Except you do - you have no control over the incoming archives on the classpath, you're not just normalizing the jars your projects are producing. This accomodates the very real possibility of a malformed manifest or properties file in an archive that otherwise would not be a problem at all either in the build or at runtime. `ZipHasher` has an _identical_ fallback for malformed zip archives, I'm doing exactly the same thing here:

https://github.com/gradle/gradle/blob/fd00959c27bdb4e940a0066eb462c87a2868a69e/subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/ZipHasher.java#L91-L93

If the fallback was to ignore the file completely I'd agree but we fallback to the current behaviour and hash the file contents.

Reply:
To make all of this code safer it likely makes sense to make the internal exception handling far more granular so a malformed properties/manifest/zip file throws a specific exception that allows a fallback, with all other failures bubbling up. I'll look at that now.

Reply:
See https://github.com/gradle/gradle/pull/11937/commits/271389aedd8bd52be788f9b15d465061d0eafa93